typescript/token/contracts/libs/FastTransfer.sol

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.8.0;
+
+import {Message} from "./Message.sol";
+import {TokenRouter} from "./TokenRouter.sol";
+
+import {TypeCasts} from "@hyperlane-xyz/core/contracts/libs/TypeCasts.sol";
+
+/**
+ * @title Common FastTransfer functionality for ERC20 Tokens with remote transfer support.
+ * @author Abacus Works
+ */
+abstract contract FastTransfer is TokenRouter {
+    using TypeCasts for bytes32;
+    using Message for bytes;
+    /**
+     * @notice `FastTransferMetadata` is the LP data stored against `fastTransferId`.
+     */
+    struct FastTranferMetadata {
+        address filler;
+        address recipient;
+        uint256 amount;
+        uint256 fastFee;
+    }
+
+    uint256 public fastTransferId;
+    // maps `fastTransferId` to `FastTranferMetadata`.
+    mapping(bytes32 => FastTranferMetadata) filledFastTransfers;
+
+    /**
+     * @dev delegates transfer logic to `_transferTo`.
+     * @inheritdoc TokenRouter
+     */
+    function _handle(
+        uint32 _origin,
+        bytes32,
+        bytes calldata _message
+    ) internal virtual override {
+        bytes32 recipient = _message.recipient();
+        uint256 amount = _message.amount();
+        bytes calldata metadata = _message.metadata();
+        _transferTo(recipient.bytes32ToAddress(), amount, _origin, metadata);
+        emit ReceivedTransferRemote(_origin, recipient, amount);
+    }
+
+    /**
+     * @dev Transfers `_amount` of token to `_recipient`/`fastFiller` who provided LP.
+     * @dev Called by `handle` after message decoding.
+     */
+    function _transferTo(
+        address _recipient,
+        uint256 _amount,
+        uint32 _origin,
+        bytes calldata _metadata
+    ) internal virtual {
+        if (_metadata.length > 0) {
+            (uint256 _fastFee, uint256 _fastTransferId) = abi.decode(
+                _metadata,
+                (uint256, uint256)
+            );
+
+            FastTranferMetadata memory m_filledMetadata = filledFastTransfers[
+                keccak256(abi.encodePacked(_origin, _fastTransferId))
+            ];
+
+            m_filledMetadata.fastFee <= _fastFee &&
+                _recipient == m_filledMetadata.recipient &&
+                _amount == m_filledMetadata.amount
+                ? _fastTransferTo(m_filledMetadata.filler, _amount)
+                : _fastTransferTo(_recipient, _amount);
+        } else {
+            _fastTransferTo(_recipient, _amount);
+        }
```

Comment:
ditto for this
would prefer removing the metadata branching
also replace the ternary with a proper if/else

Reply:
The issue is there is no clean way to handle abi decode errors and we want the fast warp route to not be breaking and still implement the `transferRemote` function that does not encode any metadata.
Without the if statement this is not possible.

Reply:
@yorhodes Although I tried breaking it down into two functions, let me know what you think!

Reply:
> The issue is there is no clean way to handle abi decode errors and we want the fast warp route to not be breaking and still implement the `transferRemote` function that does not encode any metadata. Without the if statement this is not possible.

interesting, I see, that makes sense